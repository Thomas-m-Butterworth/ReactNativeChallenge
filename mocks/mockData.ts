import { CardProps } from "../store/data/types";

export const mockData: CardProps[] = [
  {
    from: "2024-01-20T14:00Z",
    to: "2024-01-20T14:30Z",
    intensity: {
      forecast: 45,
      actual: 42,
      index: "very-low",
    },
  },
  {
    from: "2024-01-20T14:30Z",
    to: "2024-01-20T15:00Z",
    intensity: {
      forecast: 89,
      actual: 85,
      index: "low",
    },
  },
];
