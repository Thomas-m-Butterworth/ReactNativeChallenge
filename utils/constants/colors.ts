export const COLORS = {
  background: "#ffe4b5",
  text: "#000000",
  veryLow: "#00bfff",
  low: "#3cb371",
  moderate: "#5f9ea0",
  high: "#ff6347",
  veryHigh: "#8b0000",
  actual: "#3cb371",
  forecast: "#00bfff",
};
