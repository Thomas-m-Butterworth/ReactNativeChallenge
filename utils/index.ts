import { COLORS } from "./constants";

export const formatTime = (isoString: string): string => {
  const date = new Date(isoString);
  return date.toLocaleTimeString([], { hour: "2-digit", minute: "2-digit" });
};

export const formatIndexLabel = (index: string): string => {
  const [firstWord, ...rest] = index.split("-");
  return [firstWord.charAt(0).toUpperCase() + firstWord.slice(1), ...rest].join(
    " "
  );
};

export const getIndexColor = (indexLabel: string): string => {
  const colorMap: Record<string, string> = {
    "very-low": COLORS.veryLow,
    low: COLORS.low,
    moderate: COLORS.moderate,
    high: COLORS.high,
    "very-high": COLORS.veryHigh,
  };

  return colorMap[indexLabel] || COLORS.text;
};
