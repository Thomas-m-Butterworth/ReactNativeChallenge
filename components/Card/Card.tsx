import React from "react";
import { StyleSheet, Text, View } from "react-native";
import { CardProps } from "../../store/data/types";
import { COLORS } from "../../utils/constants";
import { formatTime } from "../../utils";
import { Metrics } from "../Metrics";
import { Time } from "../Time";

export const Card = ({ card }: { card: CardProps }) => {
  return (
    <View style={styles.container} testID="CardContainer">
      <Time card={card} />
      <Metrics card={card} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    backgroundColor: COLORS.background,
    height: 150,
    borderRadius: 16,
    alignItems: "center",
    marginHorizontal: 16,
    marginVertical: 8,
    padding: 16,
  },
});
