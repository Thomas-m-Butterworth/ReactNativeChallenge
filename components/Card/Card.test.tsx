import { render } from "@testing-library/react-native";

import * as stories from "./Card.stories";
import { formatIndexLabel, formatTime } from "../../utils";
import { Card } from "./Card";
import { mockData } from "../../mocks/mockData";
import { COLORS } from "../../utils/constants";

describe("Card", () => {
  it("should display correct time range", () => {
    const card = mockData[0];
    const { getByTestId } = render(<Card card={card} />);

    const timeRange = getByTestId("TimeRangeContainer");
    const expectedString = `${formatTime(card.from)} - ${formatTime(card.to)}`;
    expect(timeRange).toHaveTextContent(expectedString);
  });

  it("should display the correct forecast, actual, and index values", () => {
    const card = mockData[0];
    const { getByTestId } = render(<Card card={card} />);

    const forecastText = getByTestId("ForecastText");
    expect(forecastText).toHaveTextContent(card.intensity.forecast.toString());

    const actualText = getByTestId("ActualText");
    expect(actualText).toHaveTextContent(card.intensity.actual.toString());

    const indexText = getByTestId("IndexText");
    expect(indexText).toHaveTextContent(formatIndexLabel(card.intensity.index));
  });

  it("should display expected icons", () => {
    const card = mockData[0];
    const { getByTestId } = render(<Card card={card} />);

    const forecastIcon = getByTestId("ForecastIcon");
    const actualIcon = getByTestId("ActualIcon");
    const indexIcon = getByTestId("IndexIcon");

    expect(forecastIcon).toBeTruthy();
    expect(actualIcon).toBeTruthy();
    expect(indexIcon).toBeTruthy();
  });

  it("should apply the correct color to the index icon", () => {
    const card = mockData[0];
    const { getByTestId } = render(<Card card={card} />);

    const indexIcon = getByTestId("IndexIcon");
    expect(indexIcon.props.color).toBe(COLORS.veryLow);
  });

  it("should display the index label with proper formatting", () => {
    const card = mockData[1];
    const { getByTestId } = render(<Card card={card} />);

    const indexLabel = getByTestId("IndexText");
    expect(indexLabel).toHaveTextContent("Low");
  });

  it("renders correctly", () => {
    const card = mockData[0];
    const { toJSON } = render(<Card card={card} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
