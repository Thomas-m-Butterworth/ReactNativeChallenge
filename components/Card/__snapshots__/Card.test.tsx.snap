// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#ffe4b5",
      "borderRadius": 16,
      "height": 150,
      "marginHorizontal": 16,
      "marginVertical": 8,
      "padding": 16,
    }
  }
  testID="CardContainer"
>
  <View
    style={
      {
        "alignItems": "center",
        "marginBottom": 16,
      }
    }
    testID="TimeRangeContainer"
  >
    <Text
      style={
        {
          "color": "#000000",
          "fontSize": 20,
          "fontWeight": "700",
          "letterSpacing": -0.35,
        }
      }
    >
      14:00
       - 
      14:30
    </Text>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-around",
        "marginTop": 5,
      }
    }
    testID="MetricsContainer"
  >
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
      testID="ForecastContainer"
    >
      <FeatherMock
        color="#00bfff"
        name="trending-up"
        size={30}
        testID="ForecastIcon"
      />
      <Text
        style={
          {
            "color": "#333",
            "fontSize": 18,
            "fontWeight": "500",
            "marginTop": 8,
          }
        }
        testID="ForecastText"
      >
        45
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
      testID="ActualContainer"
    >
      <FeatherMock
        color="#3cb371"
        name="eye"
        size={30}
        testID="ActualIcon"
      />
      <Text
        style={
          {
            "color": "#333",
            "fontSize": 18,
            "fontWeight": "500",
            "marginTop": 8,
          }
        }
        testID="ActualText"
      >
        42
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
      testID="IndexContainer"
    >
      <FeatherMock
        color="#00bfff"
        name="bar-chart-2"
        size={30}
        testID="IndexIcon"
      />
      <Text
        style={
          {
            "color": "#333",
            "fontSize": 18,
            "fontWeight": "500",
            "marginTop": 8,
          }
        }
        testID="IndexText"
      >
        Very low
      </Text>
    </View>
  </View>
</View>
`;
